broadcaster/backlog/backlog.go

```diff
@@ -0,0 +1,399 @@
+	if !ok {
+		return nil, fmt.Errorf("error finding backlog segment containing message with SequenceNumber %d", i)
+	}
+
+	return segment, nil
+}
+
+// Count returns the number of messages stored within the backlog.
+func (s *backlog) Count() uint64 {
+	return s.messageCount.Load()
+}
+
+// reset removes all segments from the backlog.
+func (b *backlog) reset() {
+	b.lookupLock.Lock()
+	defer b.lookupLock.Unlock()
+	b.head = atomic.Pointer[backlogSegment]{}
+	b.tail = atomic.Pointer[backlogSegment]{}
+	b.lookupByIndex = map[uint64]*backlogSegment{}
+	b.messageCount.Store(0)
+}
+
+// BacklogSegment defines the interface for backlogSegment.
+type BacklogSegment interface {
+	Start() uint64
+	End() uint64
+	Next() BacklogSegment
+	Contains(uint64) bool
+	Messages() []*m.BroadcastFeedMessage
+}
+
+// backlogSegment stores messages up to a limit defined by the backlog. It also
+// points to the next backlogSegment in the list.
+type backlogSegment struct {
+	start           atomic.Uint64
+	end             atomic.Uint64
+	messagesLock    sync.RWMutex
+	messages        []*m.BroadcastFeedMessage
+	messageCount    atomic.Uint64
+	nextSegment     atomic.Pointer[backlogSegment]
+	previousSegment atomic.Pointer[backlogSegment]
+}
+
+// newBacklogSegment creates a backlogSegment object with an empty slice of
+// messages. It does not return an interface as it is only used inside the
+// backlog library.
+func newBacklogSegment() *backlogSegment {
+	return &backlogSegment{
+		messages: []*m.BroadcastFeedMessage{},
+	}
+}
+
+// IsBacklogSegmentNil uses the internal backlogSegment type to check if a
+// variable of type BacklogSegment is nil or not. Comparing whether an
+// interface is nil directly will not work.
+func IsBacklogSegmentNil(segment BacklogSegment) bool {
+	return segment.(*backlogSegment) == nil
+}
+
+// Start returns the first message index within the backlogSegment.
+func (s *backlogSegment) Start() uint64 {
+	return uint64(s.start.Load())
+}
+
+// End returns the last message index within the backlogSegment.
+func (s *backlogSegment) End() uint64 {
+	return uint64(s.end.Load())
+}
+
+// Next returns the next backlogSegment.
+func (s *backlogSegment) Next() BacklogSegment {
+	return s.nextSegment.Load()
+}
+
+// Messages returns all of the messages stored in the backlogSegment.
+func (s *backlogSegment) Messages() []*m.BroadcastFeedMessage {
+	s.messagesLock.RLock()
+	defer s.messagesLock.RUnlock()
+	return s.messages
+}
+
+// get reads messages from the given start to end message index.
+func (s *backlogSegment) get(start, end uint64) ([]*m.BroadcastFeedMessage, error) {
+	noMsgs := []*m.BroadcastFeedMessage{}
+	if start < s.start.Load() {
+		return noMsgs, errOutOfBounds
+	}
+
+	if end > s.end.Load() {
+		return noMsgs, errOutOfBounds
+	}
+
+	startIndex := start - s.start.Load()
+	endIndex := end - s.start.Load() + 1
+
+	s.messagesLock.RLock()
+	defer s.messagesLock.RUnlock()
+	return s.messages[startIndex:endIndex], nil
```

Comment:
I think this is safe in the current implementation, but this is dangerous. It returns a reference to the underlying messages array that stays alive after the messagesLock is released. We should consider returning a copy of this slice for safety.